include/deal.II/lac/solver_gmres.h

```diff
@@ -710,81 +723,319 @@ SolverGMRES<VectorType>::givens_rotation(Vector<double> &h,
-  const boost::signals2::signal<void(int)> &reorthogonalize_signal)
+namespace internal
 {
-  Assert(dim > 0, ExcInternalError());
-  const unsigned int inner_iteration = dim - 1;
+  namespace SolverGMRESImplementation
+  {
+    template <class VectorType>
+    void
+    dot_and_add(
+      const unsigned int dim,
+      const VectorType & vv,
+      const internal::SolverGMRESImplementation::TmpVectors<VectorType>
+        &             orthogonal_vectors,
+      Vector<double> &h)
+    {
+      for (unsigned int i = 0; i < dim; ++i)
+        h[i] += vv * orthogonal_vectors[i];
+    }
 
-  // need initial norm for detection of re-orthogonalization, see below
-  double     norm_vv_start = 0;
-  const bool consider_reorthogonalize =
-    (reorthogonalize == false) && (inner_iteration % 5 == 4);
-  if (consider_reorthogonalize)
-    norm_vv_start = vv.l2_norm();
 
-  for (unsigned int i = 0; i < dim; ++i)
-    h[i] = 0;
 
-  for (unsigned int c = 0; c < 2; ++c) // 0: orthogonalize, 1: reorthogonalize
+    template <class Number>
+    void
+    dot_and_add(
+      const unsigned int                                                   dim,
+      const LinearAlgebra::distributed::Vector<Number, MemorySpace::Host> &vv,
+      const internal::SolverGMRESImplementation::TmpVectors<
+        LinearAlgebra::distributed::Vector<Number, MemorySpace::Host>>
+        &             orthogonal_vectors,
+      Vector<double> &h)
     {
-      // Orthogonalization
-      double htmp = vv * orthogonal_vectors[0];
-      h(0) += htmp;
+      AlignedVector<VectorizedArray<double>> hs(dim,
+                                                VectorizedArray<double>(0.0));
+
+      unsigned int j = 0;
+      for (unsigned int c = 0;
+           j < vv.locally_owned_size() / VectorizedArray<double>::size();
+           ++c, j += VectorizedArray<double>::size())
+        for (unsigned int i = 0; i < dim; ++i)
+          DEAL_II_OPENMP_SIMD_PRAGMA
+      for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+        hs[i][v] +=
+          orthogonal_vectors[i].local_element(j + v) * vv.local_element(j + v);
+
+      for (unsigned int i = 0; i < dim; ++i)
+        for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+          h(i) += hs[i][v];
+
+      for (; j < vv.locally_owned_size(); ++j)
+        for (unsigned int i = 0; i < dim; ++i)
+          h(i) += orthogonal_vectors[i].local_element(j) * vv.local_element(j);
+
+      Utilities::MPI::sum(h, MPI_COMM_WORLD, h);
+    }
+
+
+
+    template <class VectorType>
+    double
+    substract_and_norm(
+      const unsigned int dim,
+      const internal::SolverGMRESImplementation::TmpVectors<VectorType>
+        &                   orthogonal_vectors,
+      const Vector<double> &h,
+      VectorType &          vv)
+    {
+      for (unsigned int i = 0; i < dim; ++i)
+        vv.add(-h(i), orthogonal_vectors[i]);
+
+      return vv.l2_norm();
```

Comment:
Not sure if we want to optimize this, but the last loop iteration and subsequent norm computation could be replaced by `VectorType::add_and_dot`.